mypy/checker.py

```diff
@@ -980,7 +980,9 @@ def _visit_func_def(self, defn: FuncDef) -> None:
                         # Trying to redefine something like partial empty list as function.
                         self.fail(message_registry.INCOMPATIBLE_REDEFINITION, defn)
                 else:
-                    # TODO: Update conditional type binder.
+                    name_expr = NameExpr(defn.name)
```

Comment:
My rule of thumb is: when working with `FuncDef`, always check for `@overload` as well.

Do we need such a check here?

Reply:
Hmm, possibly. You're talking about the check on L954 right? I think it's a larger / not fully related change because currently `original_def` cannot be `OverloadedFuncDef`... so if it's a bug we need to fix semanal etc too

Reply:
Sorry for the bad phrasing :)

I was asking more about testing something like:

```python
from typing import Callable, Optional, overload

def g() -> None:
    foo: Optional[_OverloadedProtocolThatWeNeedToDefine] = None
    if foo is None:
        @overload
        def foo() -> None: ...
        @overload 
        def foo(arg: int) -> None: ...
    foo()
    foo(1)
```

Reply:
Oh I see, I should have something in `TypeChecker._visit_overloaded_func_def`... it looks like your test case will currently need a lot more work. We allow function definitions to be conditional in semanal but not overload function definitions https://github.com/python/mypy/blob/57de8dbfaef646bf6f962377f5e1a6ec1b47889f/mypy/semanal.py#L781